languages/julia/exercises/concept/annalyns-infiltration2/runtests.jl

```diff
@@ -0,0 +1,137 @@
+    @test eltype(fogs) == Bool
+    @test !all(fogs)
+    @test any(fogs)
+    mock(rand) do rnd
+        is_foggy()
+        @test called_once_with(rnd, Bool)
+    end
+end
+
+
+@testset "spying" begin
+    character_state_combinations = [
+        (false, false, false, false),
+        (false, false, true, true),
+        (false, true, false, true),
+        (false, true, true, true),
+        (true, false, false, true),
+        (true, false, true, true),
+        (true, true, true, true),
+    ]
+
+    @testset "no fog" begin
+        mock((is_foggy) => Mock(false)) do _
+            for state in character_state_combinations
+                @test can_spy(state[1:3]...) == state[4]
+            end
+        end
+    end
+
+    @testset "fog" begin
+        mock((is_foggy) => Mock(true)) do _
+            for state in character_state_combinations
+                @test !can_spy(state[1:3]...)
+            end
+        end
+    end
+end
+
+@testset "signaling prisoner" begin
+    character_state_combinations = [
+        (false, false, false),
+        (false, true, true),
+        (true, false, false),
+        (true, true, false),
+    ]
+
+    for state in character_state_combinations
+        @test can_signal_prisoner(state[1:2]...) == state[3]
+    end
+end
+
+@testset "distracted dog" begin
+    # This exercise isn't about mathematical methods to determine the distribution of a random sample,
+    # therefore we only test that is_dog_distracted() returns boolean values and not all return values are identical.
+    distractions = collect(is_dog_distracted() for _ in 1:1000)
+    @test eltype(distractions) == Bool
+    @test !all(distractions)
+    @test any(distractions)
+end
+
+@testset "freeing prisoner" begin
+    character_state_combinations_with_dog = [
+        (false, false, false, false, false),
+        (false, false, false, true, true),
+        (false, false, true, false, true),
+        (false, false, true, true, true),
+        (false, true, false, false, false),
+        (false, true, false, true, false),
+        (false, true, true, false, false),
+        (false, true, true, true, false),
+        (true, false, false, false, false),
+        (true, false, false, true, true),
+        (true, false, true, false, false),
+        (true, false, true, true, true),
+        (true, true, false, false, false),
+        (true, true, false, true, false),
+        (true, true, true, false, false),
+        (true, true, true, true, false),
+    ]
+
+    @testset "dog is not distracted" begin
+        mock((is_dog_distracted) => Mock(false)) do _
+            for state in character_state_combinations_with_dog
+                @test can_free_prisoner(state[1:4]...) == state[5]
+            end
+        end
+    end
+
+    character_state_combinations_without_dog = [
+        (false, false, false, false, false),
+        (false, false, false, true, false), # this situation is solely determined by the dog's presence
```

Comment:
To avoid the mocks, you can just exclude these three cases from the main array and test them separately, doing the same thing as for is_foggy: just check that over a large number of runs they return both true and false.

The other option in both circumstances is to hoist the random functions out of the `can_*` functions and instead pass the results in, but then why include the old functions at all?

Reply:
> To avoid the mocks, you can just exclude these three cases from the main array and test them separately, doing the same thing as for is_foggy: just check that over a large number of runs they return both true and false.

That's a good idea. I'll do that. The `called_once_with` checks could then be moved to an eventual analyzer and we can drop the dependency.

Reply:
> The other option in both circumstances is to hoist the random functions out of the can_* functions and instead pass the results in, but then why include the old functions at all?

I think there's some justification for factoring out the randomness functions to allow more thorough testing. This would allow one to test the properties of the RNG, like its distribution, separately from the logic that depends on the result.

Reply:
I like this separation of concerns, as that means tasks will become smaller which is usually a better experience for the student.

Reply:
I also think that it's a good idea to drop the `can_*` functions from this exercise. Instead we could say that the game designer wants to add some random elements to the game and asks for the functions as described.

But it's your baby, Sascha, your decision, either way will be okay :)

Reply:
Do you think we should add more random functions in that case?

I think a solution that is only

```julia
is_foggy() = rand(Bool)
loot() = rand(1:13)
is_dog_distracted() = rand() < 0.25
```

is not really enough to be interesting on its own, given that those would like be examples in the intro doc already.

---

> But it's your baby, Sascha, your decision, either way will be okay :)

I have a third exercise in this series in mind but it would work with either approach.

Reply:
Maybe something like

```julia
crate = ["Dagger", "Sword", ...]
loot(crate) = rand(crate)
```

with a new pre-req "collections".

Reply:
The concept exercises are supposed to be simple, so I think the three or four line version is fine.

Have we introduced ranges by this point? We could explain them along with other collections.

Reply:
> Have we introduced ranges by this point? 

I think randomness is not an essential concept, so we can place it wherever, really. Adding collections and perhaps loops as pre-req would be fine with me.

Reply:
We can teach them here as well. Revising material is good.